File: CS/Q142554/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList.Nodes.Operations;

namespace Q142554 {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        DataTable CreateTable()
        {
            DataTable table = new DataTable();
            DataRow dataRow;
            table.Columns.Add("ID", typeof(System.Int32));
            table.Columns.Add("ID_2", typeof(System.Int32));
            table.Columns.Add("Name", typeof(System.String));
            table.Columns.Add("Priority", typeof(System.String));
            dataRow = table.NewRow();
            dataRow["ID"] = 1;
            dataRow["ID_2"] = 0;
            dataRow["Name"] = "Project A";
            dataRow["Priority"] = "High";
            table.Rows.Add(dataRow);

            dataRow = table.NewRow();
            dataRow["ID"] = 2;
            dataRow["ID_2"] = 1;
            dataRow["Name"] = "Deliverable 1";
            dataRow["Priority"] = "Normal";
            table.Rows.Add(dataRow);

            dataRow = table.NewRow();
            dataRow["ID"] = 3;
            dataRow["ID_2"] = 2;
            dataRow["Name"] = "This task is mine A";
            dataRow["Priority"] = "High";
            table.Rows.Add(dataRow);

            dataRow = table.NewRow();
            dataRow["ID"] = 4;
            dataRow["ID_2"] = 2;
            dataRow["Name"] = "This task isn't mine";
            dataRow["Priority"] = "Low";
            table.Rows.Add(dataRow);

            dataRow = table.NewRow();
            dataRow["ID"] = 5;
            dataRow["ID_2"] = 0;
            dataRow["Name"] = "Project B";
            dataRow["Priority"] = "Normal";
            table.Rows.Add(dataRow);

            dataRow = table.NewRow();
            dataRow["ID"] = 6;
            dataRow["ID_2"] = 5;
            dataRow["Name"] = "This task is mine B";
            dataRow["Priority"] = "High";
            table.Rows.Add(dataRow);

            return table;
        }
        private void Form1_Load(object sender, EventArgs e) {
            treeList1.DataSource = CreateTable();
        }

        private void OnTreeListBeforeCheckNode(object sender, DevExpress.XtraTreeList.CheckNodeEventArgs e) {
            e.State = (e.PrevState == CheckState.Checked ? CheckState.Unchecked : CheckState.Checked);
        }

        private void OnTreeListAfterCheckNode(object sender, DevExpress.XtraTreeList.NodeEventArgs e) {
            SetCheckedChildNodes(e.Node, e.Node.CheckState);
            SetCheckedParentNodes(e.Node, e.Node.CheckState);
        }

        private void SetCheckedChildNodes(TreeListNode node, CheckState check) {
            for (int i = 0; i < node.Nodes.Count; i++) {
                node.Nodes[i].CheckState = check;
                SetCheckedChildNodes(node.Nodes[i], check);
            }
        }
        private void SetCheckedParentNodes(TreeListNode node, CheckState check) {
            if (node.ParentNode != null) {
                bool b = false;
                CheckState state;
                for (int i = 0; i < node.ParentNode.Nodes.Count; i++) {
                    state = (CheckState)node.ParentNode.Nodes[i].CheckState;
                    if (!check.Equals(state)) {
                        b = !b;
                        break;
                    }
                }
                node.ParentNode.CheckState = b ? CheckState.Indeterminate : check;
                SetCheckedParentNodes(node.ParentNode, check);
            }
        }

        private void simpleButton1_Click(object sender, EventArgs e) {
            CalcCheckedNodesOperation operation = new CalcCheckedNodesOperation();
            treeList1.NodesIterator.DoOperation(operation);
            UpdateList(operation);
        }

        private void simpleButton2_Click(object sender, EventArgs e) {
            CalcCheckedChildrenOperation operation = new CalcCheckedChildrenOperation(treeList1.FocusedNode);
            treeList1.NodesIterator.DoOperation(operation);
            UpdateList(operation);
        }

        private void UpdateList(CalcCheckedNodesOperation operation) {
            listBoxControl1.Items.Clear();
            listBoxControl1.Items.AddRange(operation.CheckedNodes.ToArray());
        }
    }

    public class CalcCheckedNodesOperation :TreeListOperation {
        private List<string> checkedNodes = new List<string>();
        public List<string> CheckedNodes { get { return checkedNodes; } }
        public override void Execute(TreeListNode node) {
            if (node.CheckState == CheckState.Checked) 
                checkedNodes.Add((string)node[node.TreeList.Columns[0]]);
        }
    }

    public class CalcCheckedChildrenOperation : CalcCheckedNodesOperation {
        private TreeListNode parent;
        private bool isCalculationStart = false;
        public CalcCheckedChildrenOperation(TreeListNode parent) { this.parent = parent; }
        public override void Execute(TreeListNode node) {
            if (node.HasAsParent(parent)) base.Execute(node);
        }
        public override bool CanContinueIteration(TreeListNode node) {
            if (isCalculationStart && !node.HasAsParent(parent)) return false;
            if (node == parent) isCalculationStart = true;
            return true;
        }
        public override bool NeedsVisitChildren(TreeListNode node) {
            return node == parent || node.HasAsParent(parent) || parent.HasAsParent(node);
        }
    }
}

File: VB/Q142554/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList.Nodes.Operations

Namespace Q142554
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Function CreateTable() As DataTable
			Dim table As New DataTable()
			Dim dataRow As DataRow
			table.Columns.Add("ID", GetType(System.Int32))
			table.Columns.Add("ID_2", GetType(System.Int32))
			table.Columns.Add("Name", GetType(System.String))
			table.Columns.Add("Priority", GetType(System.String))
			dataRow = table.NewRow()
			dataRow("ID") = 1
			dataRow("ID_2") = 0
			dataRow("Name") = "Project A"
			dataRow("Priority") = "High"
			table.Rows.Add(dataRow)

			dataRow = table.NewRow()
			dataRow("ID") = 2
			dataRow("ID_2") = 1
			dataRow("Name") = "Deliverable 1"
			dataRow("Priority") = "Normal"
			table.Rows.Add(dataRow)

			dataRow = table.NewRow()
			dataRow("ID") = 3
			dataRow("ID_2") = 2
			dataRow("Name") = "This task is mine A"
			dataRow("Priority") = "High"
			table.Rows.Add(dataRow)

			dataRow = table.NewRow()
			dataRow("ID") = 4
			dataRow("ID_2") = 2
			dataRow("Name") = "This task isn't mine"
			dataRow("Priority") = "Low"
			table.Rows.Add(dataRow)

			dataRow = table.NewRow()
			dataRow("ID") = 5
			dataRow("ID_2") = 0
			dataRow("Name") = "Project B"
			dataRow("Priority") = "Normal"
			table.Rows.Add(dataRow)

			dataRow = table.NewRow()
			dataRow("ID") = 6
			dataRow("ID_2") = 5
			dataRow("Name") = "This task is mine B"
			dataRow("Priority") = "High"
			table.Rows.Add(dataRow)

			Return table
		End Function
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
			treeList1.DataSource = CreateTable()
		End Sub

		Private Sub OnTreeListBeforeCheckNode(ByVal sender As Object, ByVal e As DevExpress.XtraTreeList.CheckNodeEventArgs) Handles treeList1.BeforeCheckNode
			e.State = (If(e.PrevState = CheckState.Checked, CheckState.Unchecked, CheckState.Checked))
		End Sub

		Private Sub OnTreeListAfterCheckNode(ByVal sender As Object, ByVal e As DevExpress.XtraTreeList.NodeEventArgs) Handles treeList1.AfterCheckNode
			SetCheckedChildNodes(e.Node, e.Node.CheckState)
			SetCheckedParentNodes(e.Node, e.Node.CheckState)
		End Sub

		Private Sub SetCheckedChildNodes(ByVal node As TreeListNode, ByVal check As CheckState)
			For i As Integer = 0 To node.Nodes.Count - 1
				node.Nodes(i).CheckState = check
				SetCheckedChildNodes(node.Nodes(i), check)
			Next i
		End Sub
		Private Sub SetCheckedParentNodes(ByVal node As TreeListNode, ByVal check As CheckState)
			If node.ParentNode IsNot Nothing Then
				Dim b As Boolean = False
				Dim state As CheckState
				For i As Integer = 0 To node.ParentNode.Nodes.Count - 1
					state = CType(node.ParentNode.Nodes(i).CheckState, CheckState)
					If Not check.Equals(state) Then
						b = Not b
						Exit For
					End If
				Next i
				node.ParentNode.CheckState = If(b, CheckState.Indeterminate, check)
				SetCheckedParentNodes(node.ParentNode, check)
			End If
		End Sub

		Private Sub simpleButton1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton1.Click
			Dim operation As New CalcCheckedNodesOperation()
			treeList1.NodesIterator.DoOperation(operation)
			UpdateList(operation)
		End Sub

		Private Sub simpleButton2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton2.Click
			Dim operation As New CalcCheckedChildrenOperation(treeList1.FocusedNode)
			treeList1.NodesIterator.DoOperation(operation)
			UpdateList(operation)
		End Sub

		Private Sub UpdateList(ByVal operation As CalcCheckedNodesOperation)
			listBoxControl1.Items.Clear()
			listBoxControl1.Items.AddRange(operation.CheckedNodes.ToArray())
		End Sub
	End Class

	Public Class CalcCheckedNodesOperation
		Inherits TreeListOperation

'INSTANT VB NOTE: The field checkedNodes was renamed since Visual Basic does not allow fields to have the same name as other class members:
		Private checkedNodes_Renamed As New List(Of String)()
		Public ReadOnly Property CheckedNodes() As List(Of String)
			Get
				Return checkedNodes_Renamed
			End Get
		End Property
		Public Overrides Sub Execute(ByVal node As TreeListNode)
			If node.CheckState = CheckState.Checked Then
				checkedNodes_Renamed.Add(DirectCast(node(node.TreeList.Columns(0)), String))
			End If
		End Sub
	End Class

	Public Class CalcCheckedChildrenOperation
		Inherits CalcCheckedNodesOperation

		Private parent As TreeListNode
		Private isCalculationStart As Boolean = False
		Public Sub New(ByVal parent As TreeListNode)
			Me.parent = parent
		End Sub
		Public Overrides Sub Execute(ByVal node As TreeListNode)
			If node.HasAsParent(parent) Then
				MyBase.Execute(node)
			End If
		End Sub
		Public Overrides Function CanContinueIteration(ByVal node As TreeListNode) As Boolean
			If isCalculationStart AndAlso Not node.HasAsParent(parent) Then
				Return False
			End If
			If node Is parent Then
				isCalculationStart = True
			End If
			Return True
		End Function
		Public Overrides Function NeedsVisitChildren(ByVal node As TreeListNode) As Boolean
			Return node Is parent OrElse node.HasAsParent(parent) OrElse parent.HasAsParent(node)
		End Function
	End Class
End Namespace